### DISCUSSION TITLE
How to deal with `PHPStan\ShouldNotHappenException: Internal error`?

### DISCUSSION THREAD

**Discussion by Wirone** [rectorphp/rector#6743] [Q&A] [answered]
We're on Rector `0.11.56` and currently we can't upgrade to newer version because tests for our custom Rector rules fail in random way (I run and I get 2 errors, I run again and I get 9...). Output does not help:

```
5) Utils\Rector\Tests\Rule\Naming\HungarianNotationRector\HungarianNotationRectorTest::test with data set #4 (RectorPrefix20211013\Symplify\SmartFileSystem\SmartFileInfo Object (...))
PHPStan\ShouldNotHappenException: Internal error.

phar:///app/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/ScopeContext.php:36
phar:///app/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:1900
/app/vendor/rector/rector/packages/NodeTypeResolver/PHPStan/Scope/PHPStanNodeScopeResolver.php:235
/app/vendor/rector/rector/packages/NodeTypeResolver/PHPStan/Scope/PHPStanNodeScopeResolver.php:131
phar:///app/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:317
phar:///app/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:259
phar:///app/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:474
phar:///app/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:227
/app/vendor/rector/rector/packages/NodeTypeResolver/PHPStan/Scope/PHPStanNodeScopeResolver.php:155
/app/vendor/rector/rector/packages/NodeTypeResolver/PHPStan/Scope/PHPStanNodeScopeResolver.php:143
/app/vendor/rector/rector/packages/NodeTypeResolver/NodeScopeAndMetadataDecorator.php:81
/app/vendor/rector/rector/src/Application/FileProcessor.php:49
/app/vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php:80
/app/vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php:144
/app/vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php:81
/app/vendor/rector/rector/src/Application/ApplicationFileProcessor.php:76
/app/vendor/rector/rector/src/Application/ApplicationFileProcessor.php:57
/app/vendor/rector/rector/packages/Testing/PHPUnit/AbstractRectorTestCase.php:149
/app/vendor/rector/rector/packages/Testing/PHPUnit/AbstractRectorTestCase.php:114
/app/vendor/rector/rector/packages/Testing/PHPUnit/AbstractRectorTestCase.php:105
/app/utils/rector/tests/Rule/Naming/HungarianNotationRector/HungarianNotationRectorTest.php:18
```

How can I debug it efficiently, so I will find actual problem? Because right now I don't know if it's our rule's problem or Rector's one.

**Comment by TomasVotruba** [comment, marked as answer]
Hi, it's hard to say where the issue is. Could you create reproductible repository?

**Reply by Wirone** [reply]
Probably I could, but the thing is I would like to learn how to deal with it and find actual problems on my own, so I can fix it if the problem is on our side or create issue with proper info instead of preparing reproducer repo (our app is really large and I don't know what should I migrate to such repo, also most of things I just can't publish).

**Reply by TomasVotruba** [reply]
I see. You ask for debugging tips :)

This is not related to Rector, but rather to any stack trace. When I get some, I always try to dump the line one by one, till I found something meaninful.

```bash
// why its thrown here?
phar:///app/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/ScopeContext.php:36
phar:///app/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:1900

// what values are on this line at the time the exception happens?
/app/vendor/rector/rector/packages/NodeTypeResolver/PHPStan/Scope/PHPStanNodeScopeResolver.php:235
```

**Reply by Wirone** [reply]
Yeah, I know how to debug 😉 I rather was asking about some internal tricks (like Rector config options, PHPUnit config, modifying Rector's code) that you personally use.  

It would be helpful if there would be fixture name printed instead of `with data set #4` (like `with fixture 'fix_method_arguments.php.inc'`). After successful check output contains diff and full path to the processed fixture, but in error output we don't know which fixtures is causing it.

FYI: I've updated to latest build `84e32824` and still I have `ShouldNotHappenException`s. The weirdest thing is that I have 0 errors when I modify `HungarianNotationRectorTest::provideData()` and test each fixture one by one... Then I get 2 or 9 errors alternately when I run all 12 fixtures, without changing the code.. Cleared PHPUnit result cache etc, no fix yet 😩

**Reply by TomasVotruba** [reply]
Nothing special, I keep it simple :).

> It would be helpful if there would be fixture name printed instead of with data set #4 (like with fixture 'fix_method_arguments.php.inc'). 

If you can keep number filtering like for phpunit `vendor/bin/phpunit --filter test#44`, go for PR that allows this :+1:

**Reply by Wirone** [reply]
Ahh, this output is PHPUnit's 🤦 Nevermind then.

Any ideas why all fixtures are correct when running standalone (or using `--filter`), but fail when running together?

**Reply by TomasVotruba** [reply]
There is some state leftover in the rule or some service.

**Reply by TomasVotruba** [reply]
What exact errors did you found out by debugging these 3 lines?
https://github.com/rectorphp/rector/discussions/6743#discussioncomment-1472910

**Reply by Wirone** [reply]
Unfortunately I did not have time for digging deeper yet, I'll try to do it in upcoming days. Will get back as soon as I check it. In the meantime I try nightly releases, maybe it will be fixed magically 😉 The problem was introduced in 0.11.57 for sure, because this is first version when our tests fail.

**Reply by TomasVotruba** [reply]
I see. It will be in here then: https://github.com/rectorphp/rector-src/compare/0.11.56...0.11.57

Possibly the trait-related PR with node scope resolver: https://github.com/rectorphp/rector-src/pull/952

**Reply by TomasVotruba** [reply]
Most likely this file change: https://github.com/rectorphp/rector-src/pull/952/files#diff-625e519d3d496a50287cab05a2ffb79a95877d1850083f9a90fdd74958a01020

**Reply by Wirone** [reply]
@TomasVotruba I think it's a good path, when I removed 2 fixtures for traits the other fixtures were tested properly on `dev-main#84e32824ab48adc4b4489f3accc5f27dbc22d4a4` 👍 I'll try to find what's the problem with those fixtures. Thank you!

In the meantime I'll leave here those fixtures, maybe you will find out what's the problem:

```
<?php

namespace Utils\Rector\Tests\Rule\Naming\HungarianNotationRector\Fixture;

trait Foo
{
    abstract public function bar($pBar, $mBaz);
}

?>
-----
<?php

namespace Utils\Rector\Tests\Rule\Naming\HungarianNotationRector\Fixture;

trait Foo
{
    abstract public function bar($bar, $baz);
}

?>
```

and

```
<?php

namespace Utils\Rector\Tests\Rule\Naming\HungarianNotationRector\Fixture;

// No changes should be made here by Rector - we can't properly determine trait usage and name conflicts for properties

trait Foo
{
    private $mFoo = 0;
    private $pBar = 1;
}
```

**Reply by TomasVotruba** [reply]
At first sight, the traits should have unique name. Now both files have the same name.

Sometimes it makes bugs, as the container in one test is build just once.

**Reply by Wirone** [reply]
@TomasVotruba Bingo! After renaming classes, interfaces and traits to unique names within namespace (fixtures for the same rule) it passes correctly! 😁 Thank you very much 🍻 

PS. That weird tests results (2 errors, 9 errors, 2 errors and so on) probably were caused by `executionOrder="defects"` in PHPUnit's config. So every other run was in the opposite order which was causing different collisions of identifiers).

**Reply by TomasVotruba** [reply]
Great, I'm glad it works :+1: :clap:
